src/main/io/osd_elements_impl.h

```diff
@@ -0,0 +1,918 @@
+    if (featureIsEnabled(FEATURE_ESC_SENSOR) && osdWarnGetState(OSD_WARNING_ESC_FAIL)) {
+        char escWarningMsg[OSD_FORMAT_MESSAGE_BUFFER_SIZE];
+        unsigned pos = 0;
+
+        const char *title = "ESC";
+
+        // center justify message
+        while (pos < (OSD_WARNINGS_MAX_SIZE - (strlen(title) + getMotorCount())) / 2) {
+            escWarningMsg[pos++] = ' ';
+        }
+
+        strcpy(escWarningMsg + pos, title);
+        pos += strlen(title);
+
+        unsigned i = 0;
+        unsigned escWarningCount = 0;
+        while (i < getMotorCount() && pos < OSD_FORMAT_MESSAGE_BUFFER_SIZE - 1) {
+            escSensorData_t *escData = getEscSensorData(i);
+            const char motorNumber = '1' + i;
+            // if everything is OK just display motor number else R, T or C
+            char warnFlag = motorNumber;
+            if (ARMING_FLAG(ARMED) && osdConfig()->esc_rpm_alarm != ESC_RPM_ALARM_OFF && calcEscRpm(escData->rpm) <= osdConfig()->esc_rpm_alarm) {
+                warnFlag = 'R';
+            }
+            if (osdConfig()->esc_temp_alarm != ESC_TEMP_ALARM_OFF && escData->temperature >= osdConfig()->esc_temp_alarm) {
+                warnFlag = 'T';
+            }
+            if (ARMING_FLAG(ARMED) && osdConfig()->esc_current_alarm != ESC_CURRENT_ALARM_OFF && escData->current >= osdConfig()->esc_current_alarm) {
+                warnFlag = 'C';
+            }
+
+            escWarningMsg[pos++] = warnFlag;
+
+            if (warnFlag != motorNumber) {
+                escWarningCount++;
+            }
+
+            i++;
+        }
+
+        escWarningMsg[pos] = '\0';
+
+        if (escWarningCount > 0) {
+            osdFormatMessage(element->buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, escWarningMsg);
+            SET_BLINK(OSD_WARNINGS);
+            return;
+        }
+    }
+#endif // USE_ESC_SENSOR
+
+    if (osdWarnGetState(OSD_WARNING_BATTERY_WARNING) && batteryState == BATTERY_WARNING) {
+        osdFormatMessage(element->buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, "LOW BATTERY");
+        SET_BLINK(OSD_WARNINGS);
+        return;
+    }
+
+#ifdef USE_RC_SMOOTHING_FILTER
+    // Show warning if rc smoothing hasn't initialized the filters
+    if (osdWarnGetState(OSD_WARNING_RC_SMOOTHING) && ARMING_FLAG(ARMED) && !rcSmoothingInitializationComplete()) {
+        osdFormatMessage(element->buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, "RCSMOOTHING");
+        SET_BLINK(OSD_WARNINGS);
+        return;
+    }
+#endif // USE_RC_SMOOTHING_FILTER
+
+    // Show warning if battery is not fresh
+    if (osdWarnGetState(OSD_WARNING_BATTERY_NOT_FULL) && !ARMING_FLAG(WAS_EVER_ARMED) && (getBatteryState() == BATTERY_OK)
+          && getBatteryAverageCellVoltage() < batteryConfig()->vbatfullcellvoltage) {
+        osdFormatMessage(element->buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, "BATT < FULL");
+        return;
+    }
+
+    // Visual beeper
+    if (osdWarnGetState(OSD_WARNING_VISUAL_BEEPER) && isBeeperOn()) {
```

Comment:
This will have to be made latching again (see how `showVisualBeeper` was used) - otherwise the interference between OSD updates and beeper on / off cycles will result mean the warning is only ever flickering.

Reply:
There was no "latching", that's why I removed it. The old logic simply set the flag based on beeper state every refresh cycle and it was cleared after drawing it.
```
void osdUpdate(timeUs_t currentTimeUs)
{
    static uint32_t counter = 0;

    if (isBeeperOn()) {
        showVisualBeeper = true;
    }
```
and then farther down in the same function
```
    if (counter % DRAW_FREQ_DENOM == 0) {
        osdRefresh(currentTimeUs);
        showVisualBeeper = false;
    } else {
        // rest of time redraw screen 10 chars per idle so it doesn't lock the main idle
        displayDrawScreen(osdDisplayPort);
    }
```
Then the logic to actually draw the element:
```
            // Visual beeper
            if (osdWarnGetState(OSD_WARNING_VISUAL_BEEPER) && showVisualBeeper) {
                osdFormatMessage(buff, OSD_FORMAT_MESSAGE_BUFFER_SIZE, "  * * * *");
                break;
            }
```
So what's going on is that the `osdUpdate()` function is called by the OSD task at 60hz. Each call is converted into 5 phases. The first phase clears the screen and draws all the elements. The remaining 4 phases are used to stream out any updates in the screenBuffer to the max7456 (100 character positions per each phase).

So from the element drawing perspective the visual beeper warning element is just based on the current state of `isBeeperOn()`. 

Latching isn't really necessary anyway as the entire screenBuffer is cleared every update and only the active elements redrawn. So if the beeper is on it will be drawn, it not then it will be skipped and the position left blank from the erase. The actual OSD display is refreshed at 12hz (60 / 5) so it tracks the state of the beeper pretty well (except for really fast beeps). Nothing has changed in that regard.

The only reason to add some actual latching of minimum on time would be to make sure that a visual indication occurs for every beep even if very short. This would be a change to the visual beeper handling in general - not related to changes in this PR. The downside to this is that depending on the time constant chosen for the "latching" the display won't track as well with the audible beeper for short duration beeps.

Reply:
Have you tested it? The warning was flickering before latching was added, and latching fixed that.

The critical point in the code is the `if (counter % DRAW_FREQ_DENOM == 0)` - the latching adds a hold until the draw cycle has been completed, meaning that if the beeper is on during any part of the 12 Hz cycle the warning will be shown.

Reply:
Yes, I've tested quite extensively and there's no discernible difference in old vs. new. The only risk without the latching is if the beeper comes on and turns off within the final 4/5 of the cycle it won't displayed. But we're talking about a 66.7ms period. Even with the previous logic the best case in this scenario is that the visual beeper would be displayed for 1 OSD refresh cycle or 83ms.

Reply:
I'm not really sure it's necessary (why I removed it) but I don't have a problem with adding back the latching if we think it's important. The lesser used use case of OSD's that don't use the max7456 (MWOSD over serial?) have their phase count set to 10 so in that case the window for missing a beeper display is 150ms so it may make more sense in this case.

Reply:
I am not 100% sure it is necessary, I'd have to test with the current code. Way back when I added it it was necessary to make sure the warning is shown clearly and not just as a flicker, especially for sequences of short beeps, but it is possible that the way the OSD loop is executed has changed in the mean time and it is no longer necessary - will need to test.

Reply:
It's not a big deal and I've put the latching logic back in the revised version I'm finishing up. It doesn't seem to be needed based on the current code but it doesn't hurt anything. If anything it gives more flexibility if there are future changes to the task frequency or how the phases affect element drawing/max7456 streaming in the future. Some follow-on performance improvements that I mentioned above will likely do this so better safe than sorry at this point. Can reevaluate after that to see if the latching is still needed.